Add tutorial for creating HPC-like K8 app (#590)
OSC/ood-documentation: source/app-development/tutorials-interactive-apps/k8s-like-hpc-jupyter.rst

.. _app-development-tutorials-interactive-apps-k8s-like-hpc-jupyter:

Add a Jupyter App on a Kubernetes Cluster that behaves like HPC compute
=======================================================================

This tutorial will walk you through creating an interactive Jupyter app that
your users will use to launch a `Jupyter Notebook Server`_ in a kubernetes cluster.
The container will behave much like a HPC compute node. This has the benefit that
a single app can serve both traditional HPC as well as Kubernetes.

It assumes you have a working understanding of app development already. The purpose of
this document is to describe how to write apps specifically for a kubernetes cluster,
so it skips a lot of important details about app development that may be found in
other tutorials like :ref:`app-development-tutorials-interactive-apps-add-jupyter`.

> We're going to be looking at the `bc osc jupyter`_ app which you can fork, clone

Reviewer: ```suggestion
We're going to be looking at the `bc osc jupyter`_ app which is OSC's production Jupyter app. You can fork, clone
```

and modify for your site.  This page also holds the `submit yml in full`_ for reference.

The container
-------------

The container to make Kubernetes pods look like HPC compute would end up needing all
the OS packages present on the HPC environment.  The OS of the container itself would also
need to compatible with packages installed on HPC environment so if you run RHEL 8 on HPC
you would also need to run RHEL 8 inside the container.

Things like Lmod and HPC applications will need to be run inside the Pod's container just like

> if a job was spawned in the traditional HPC resource manager.

Reviewer: ```suggestion
if a job was spawned in a traditional HPC resource manager.
```

Switch between SLURM and Kubernetes
-----------------------------------

The first big change from a traditional HPC interactive app is the main YAML structure is wrapped

> in a large ``if`` statement based on the cluster choice. If a user choose one of the HPC clusters,
> the SLURM submit YAML is rendered, otherwise the Kubernetes YAML is rendered.

Reviewer: ```suggestion
in a large ``if`` statement based on the cluster choice. If a user chooses one of the SLURM clusters,
the SLURM submit YAML is rendered, otherwise the Kubernetes YAML is rendered.
```

Reviewer: Actually this sentence should be reworded to describe our environment. Notably that Owens and Pitzer are Slurm and that that our kubernetes cluster is just called 'kubernetes'.

Author: Tried to reword describing environment

Here is the beginning of the block:

.. code-block:: text

  <% if cluster =~ /owens|pitzer/ -%>
  ---
  batch_connect:
    template: "basic"
    conn_params:
      - jupyter_api
  script:
  ...SLURM specific...

Here is the logic to select Kubernetes:

.. code-block:: text

  <% elsif cluster =~ /kubernetes/
  ...Ruby variables setup here...

container spec
--------------

Let's look at this section first of the Kubernetes block.  Here you must specify the ``name``, ``image``
and ``command``.  The name determines the Pod Id (the job name in HPC parlance).

The ``image`` should be the HPC container image and ``command`` will be the job script that has been adapted to
work with both SLURM and Kubernetes.  The ``command`` will be run from the user's home directory and will cover mount

> requirements in :ref:`mount requirements <kubernetes-mount-requirements>`.

Reviewer: Don't know how to word this off hand, but: 

The `command` would be the same for every site & app and is the exact same script that would be ran if the job were ran on Slurm. Which means it runs `before.sh`, `script.sh`, `after.sh` and any other template files just like apps ran on other HPC clusters.

This point I think will help them conceptualize what's happening here. Same script, same entrypoint just totally different environment (kinda/not really after all the setup).

Author: Added a paragraph about how the script is generated using same files as traditional HPC

Next you can specify additional environment variables in ``env``. 

.. code-block:: yaml

  native:
    container:
      name: "jupyter"
      image: "docker-registry.osc.edu/ondemand/ondemand-base-rhel7:0.3.1"
      image_pull_policy: "IfNotPresent"
      command: ["/bin/bash","-l","<%= staged_root %>/job_script_content.sh"]
      restart_policy: 'OnFailure'
      env:
        NB_UID: "<%= Etc.getpwnam(ENV['USER']).uid %>"
        NB_USER: "<%= ENV['USER'] %>"
        NB_GID: "<%= Etc.getpwnam(ENV['USER']).gid %>"
        CLUSTER: "<%= compute_cluster %>"

Here is the default environment. You can use a ``null`` here to unset any of these.

.. code-block:: text

  USER: username,
  UID: run_as_user,
  HOME: home_dir,
  GROUP: group,
  GID: run_as_group,

>   KUBECONFIG: '/dev/null'

Reviewer: I don't think we need this bit, it's the same as the other.

Author: Removed

resource requests

> -----------------

Reviewer: I don't think we need this bit about resources either.

Author: Removed

.. code-block:: yaml

    container:
      # ...
      port: "8080"
      cpu_request: "0.200"
      cpu_request: "4"
      memory_request: "500Mi"
      memory_limit: "4Gi"

See `kubernetes pod memory`_ and `kubernetes pod cpu`_ for more details.

.. _kubernetes-mount-requirements:

mounts
------

For a pod to look like an HPC environment the home directory of the user and any shared filesystems would
need to be mounted on Kubernetes worker nodes and then made available to the pod.

In the example a Ruby structure is created to streamline some of the direct mounts where the path outside
the container is the same as the path inside the container:

.. code-block:: ruby

   mounts = {
     'home'    => OodSupport::User.new.home,
     'support' => OodSupport::User.new('support').home,
     'project' => '/fs/project',
     'scratch' => '/fs/scratch',
     'ess'     => '/fs/ess',
   }

These mounts are defined in the YAML using a loop:

.. code-block:: text

    mounts:
    <%- mounts.each_pair do |name, mount| -%>
      - type: host
        name: <%= name %>
        host_type: Directory
        path: <%= mount %>
        destination_path: <%= mount %>
    <%- end -%>

> Additional mounts are needed to make the pod behave like a HPC compute node. Following are mounted into the container

Reviewer: ```suggestion
Additional mounts are needed to make the pod behave like a HPC compute node. Following are mounted into the container:
```
